src/IO/ReadBufferFromS3.cpp

```diff
@@ -29,26 +31,58 @@ namespace ErrorCodes
 
 
 ReadBufferFromS3::ReadBufferFromS3(
-    std::shared_ptr<Aws::S3::S3Client> client_ptr_, const String & bucket_, const String & key_, size_t buffer_size_)
-    : SeekableReadBuffer(nullptr, 0), client_ptr(std::move(client_ptr_)), bucket(bucket_), key(key_), buffer_size(buffer_size_)
+    std::shared_ptr<Aws::S3::S3Client> client_ptr_, const String & bucket_, const String & key_, UInt64 s3_max_single_read_retries_, size_t buffer_size_)
+    : SeekableReadBuffer(nullptr, 0)
+    , client_ptr(std::move(client_ptr_))
+    , bucket(bucket_)
+    , key(key_)
+    , s3_max_single_read_retries(s3_max_single_read_retries_)
+    , buffer_size(buffer_size_)
 {
 }
 
-
 bool ReadBufferFromS3::nextImpl()
 {
-    if (!initialized)
-    {
+    /// Restoring valid value of `count()` during `nextImpl()`. See `ReadBuffer::next()`.
+    pos = working_buffer.begin();
+
+    if (!impl)
         impl = initialize();
-        initialized = true;
-    }
 
     Stopwatch watch;
-    auto res = impl->next();
+    bool next_result = false;
+
+    for (Int64 attempt = static_cast<Int64>(s3_max_single_read_retries); attempt >= 0; --attempt)
+    {
+        if (!impl)
+            impl = initialize();
+
+        try
+        {
+            next_result = impl->next();
+            /// FIXME. 1. Poco `istream` cannot read less than buffer_size or this state is being discarded during
+            ///           istream <-> iostream conversion. `gcount` always contains 0,
+            ///           that's why we always have error "Cannot read from istream at offset 0".
+
+            break;
+        }
+        catch (const Exception & e)
```

Comment:
Catching **any** exception does not looks sane, @excitoon what do you think about enumerate codes that should be retriable?